### PR TITLE
Allow training and sampling from only upsampler unet

### PR DESCRIPTION
**Pull request by Veldrovive**
It was possible to train only the upsampler before, but this pull request makes it easier and reduces ram usage by having the decoder trainer remove unets that are not being trained. This requires some changes to the trainer object to allow for the possibility of missing unets when saving. This is more convenient for combining the state dict of multiple runs each where a single unet was trained.

At sampling time, the capability to specify a start unet as well as an end unet is added. If using a start unet greater than 1 the sample image also needs to be passed in.

This is not yet ready for review, but I wanted to have it open because I found that all testing metrics and validation loss begin increasing after the first epoch in [this run](https://wandb.ai/veldrovive/dalle2_train_decoder_upsamplers/runs/2te23qxt). It is unclear to me whether this is caused by overfitting or something else.

I have a [new run](https://wandb.ai/veldrovive/dalle2_train_decoder_upsamplers/runs/1iyvavsj) with an updated version of the branch going in case some issues were already fixed.

### CODE REVIEW

**Comment by lucidrains** [comment]
@Veldrovive oh yup that doesn't look great

which version of dalle2-pytorch was the first run done with? was it before the recent commits with the extra skip connections?

**Comment by lucidrains** [comment]
so the upsampling actually depends on this gaussian blur done on the image before using it as conditioning https://github.com/lucidrains/DALLE2-pytorch/blob/main/dalle2_pytorch/dalle2_pytorch.py#L1802 i believe the paper had it at `0.6` or some random float around that

i'll check later and update this default value if i didn't set this correctly! 🙏

**Comment by Veldrovive** [comment]
@lucidrains Yes, it was before the skip connections. The original was branched off of the commit f5760bdb92fb8fe9f53aca4a87b3e4b9ef8cdbaa while the new one is up to date with current.

**Comment by lucidrains** [comment]
there is also additional BSR noising done on the low resolution conditioning, which i haven't built into the repository yet

since i'm seeing liftoff at imagen-pytorch, i may just bring in the noise augmentation technique there instead, for simplicity

**Comment by lucidrains** [comment]
> @lucidrains Yes, it was before the skip connections. The original was branched off of the commit [f5760bd](https://github.com/lucidrains/DALLE2-pytorch/commit/f5760bdb92fb8fe9f53aca4a87b3e4b9ef8cdbaa) while the new one is up to date with current.

ok cool! yeah researchers at imagen-pytorch have reported issues with the memory efficient version, which i see you have turned on, but i think the extra skip connections will help

if it doesn't, we can try upping the blur sigma and bring in extra noising

**Comment by lucidrains** [comment]
Screen Shot 2022-06-30 at 5 01 32 PM

yup appears to be 0.6, let me fix that

**Comment by Veldrovive** [comment]
Thanks for the quick fixes. I restarted the new run with the changes and I will let it run overnight. Should go through a couple epochs.

**Comment by lucidrains** [comment]
honestly not tot sure what is going on

i brought back the old upsample (convtranspose2d) since i recall someone over at imagen-pytorch telling me the upsampler was working fine for them at this repository before i made that change

**Comment by lucidrains** [comment]
also added a way to return the low resolution conditioning image https://github.com/lucidrains/DALLE2-pytorch/blob/main/dalle2_pytorch/dalle2_pytorch.py#L2252 ,  just in case that isn't generated correctly

**Comment by lucidrains** [comment]
@marunine i'm sorry to ping you like this, but did you say you tried training an upsampler from this repository, and it was successful?

**Comment by Veldrovive** [comment]
Yep, that definitely still looks wrong. I'll start a new run and have it log the low-res conditioning images as well.

**Comment by lucidrains** [comment]
@Veldrovive  @marunine Marunine identified a potential issue with the low resolution conditioning https://github.com/lucidrains/imagen-pytorch/issues/94#issuecomment-1177837279 I have put in a patch for that, just fyi!

**Comment by Veldrovive** [comment]
Results [here](https://wandb.ai/veldrovive/upsamplers/runs/25nvpkq4). At least for now the results do not look meaningfully different. I will leave the run going over night.

**Comment by jacobwjs** [comment]
The results look like some serious ringing (overshoot and undershoot) going on. Has anyone looked at activations coming out of down/up blocks in the Unet?

**Comment by rom1504** [comment]
@jacobwjs for all the wandb links we share, models are available as artifacts
if you'd like to investigate these activations it'd be appreciated

**Comment by lucidrains** [comment]
one thing I'm wondering is whether classifier free guidance should be turned off for the upsamplers

**Comment by lucidrains** [comment]
I'll start running experiments and hopefully debug this next week

**Comment by lucidrains** [comment]
may be able to dedicate some time thursday and friday to this problem and get it resolved! i think it may be the last thing

**Comment by lucidrains** [comment]
yea weird, i'm not seeing the issue training an upsampler in my other ddpm repo, but it is without classifier free guidance

**Comment by lucidrains** [comment]
just kidding, i didn't set it up correctly, rerunning the experiments sigh

**Comment by lucidrains** [comment]
yea its looking good for me, at least early in training

i'll leave it on overnight and see if those artifacts appear tomorrow morning

**Comment by lucidrains** [comment]
doh, i think i'm starting to see it locally too

ok, i'll sink another day of experiments tomorrow, figure out what's going on

**Comment by lucidrains** [comment]
looking at early results of a second experiment :eyes: i think the solution may end up being really simple (use nearest neighbor upsample for all resizing operations) will know for sure tomorrow morning

**Comment by lucidrains** [comment]
i think i figured out the problem, just have to use nearest neighbor resizing for any low resolution conditioning stuff

**Comment by lucidrains** [comment]
sample-cond-8
sample-8

**Comment by lucidrains** [comment]
but i think i'll keep the nearest neighbor change. there could potentially be multiple issues

**Comment by lucidrains** [comment]
sample-cond-10
sample-10

10k

**Comment by lucidrains** [comment]
sample-cond-11
sample-11

11k looking great!

**Comment by lucidrains** [comment]
sample-cond-22
sample-22

22k pretty sure that was it

**Comment by rom1504** [comment]
Is this PR ready? If yes i think it would be great to merge so it's not needed to rebase on main very often and resolve conflicts
I looked at the diffs and it seems ok

**Comment by rom1504** [comment]
Ah failing the test though, that should be fixed first

**Comment by rom1504** [comment]
looked a bit more, something is wrong with return_lowres_cond_image and cond_image being returned or not
I'm not sure what's the intent with this param though

**Comment by lucidrains** [comment]
I was just addressing the artifacts seen during the last training run. Should be good now!

**Comment by Veldrovive** [comment]
New [run](https://wandb.ai/veldrovive/upsampler/runs/ji74rtou) with nearest neighbor.

In terms of this pull request, it is not ready to be merged because there is testing code in place that only works when training upsamplers and errors otherwise. I could make it support the initial unet as well, but it makes more sense to me to just get upsampling to work and then remove the tests.

File: dalle2_pytorch/dalle2_pytorch.py [hunk 1]
```diff
@@ -2489,12 +2493,19 @@ def sample(
         assert not (not self.condition_on_text_encodings and exists(text_encodings)), 'decoder specified not to be conditioned on text, yet it is presented'
 
         img = None
+        if start_at_unet_number > 1:
+            # Then we are not generating the first image and one must have been passed in
+            assert exists(image), 'image must be passed in if starting at unet number > 1'
+            assert image.shape[0] == batch_size, 'image must have batch size of {} if starting at unet number > 1'.format(batch_size)
+            img = image
```
[the remark this answers is not in the document]
Comment (lucidrains) [reply dalle2_pytorch/dalle2_pytorch.py:2500]: but if the image being passed in is already sized correctly, this is a non-issue (although an extra resize just to make sure would never hurt)

**Comment by lucidrains** [comment]
after we solve this issue, this repository deserves a 1.0 lol

**Comment by lucidrains** [comment]
dealing one more blow to the problem https://github.com/lucidrains/DALLE2-pytorch/commit/a34f60962abac516a22c66bcdf0db76161754e14 seems to be converging a bit faster on my local machine

File: dalle2_pytorch/dalle2_pytorch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Veldrovive) [reply dalle2_pytorch/dalle2_pytorch.py:2500]: I was under the impression it was getting resized somewhere, but I will have to make sure of that when I get some time to dig around.

[the remark this answers is not in the document]
Comment (lucidrains) [reply dalle2_pytorch/dalle2_pytorch.py:2500]: ok take your time!

also @nousr, get a github sponsors page up! i'm going to add you both to the sponsors list

you both stepped up in a big way and i wish there were some way to reward you for an important industry-destroying project like this one other than recognition

**Comment by lucidrains** [comment]
yea, it is looking great for me locally, without classifier free guidance

**Comment by lucidrains** [comment]
sample-cond-21
sample-21

oxford flowers are looking great for me locally at 21k :rose: 

ok, i'll circle back tomorrow! feeling confident we can solve this :)

**Comment by Veldrovive** [comment]
[This run](https://wandb.ai/veldrovive/upsampler/runs/1muaeil8?workspace=user-veldrovive) has produced good results and I have removed the testing code that caused issues with single unet training.

**Comment by lucidrains** [comment]
@Veldrovive well done!
